charts/envoy/templates/deployment.yaml

```diff
@@ -0,0 +1,71 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "envoy.fullname" . }}-envoy
+  labels:
+    {{- include "envoy.labels" . | nindent 4 }}
+spec:
+  minReadySeconds: 0
+  replicas: {{ .Values.replicaCount }}
+  selector:
+    matchLabels:
+      {{- include "envoy.selectorLabels" . | nindent 6 }}
+  {{- with .Values.strategy }}
+  strategy:
+  {{- toYaml . | nindent 4 }}
+  {{- end }}
+  template:
+    metadata:
+      labels:
+        {{- include "envoy.selectorLabels" . | nindent 8 }}
+    spec:
+    {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+    {{- end }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      restartPolicy: Always
+      terminationGracePeriodSeconds: 60
+      containers:
+        - name: {{ .Chart.Name }}
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.version }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          ports:
+          - containerPort: 50051
+          - containerPort: 50052
+          - containerPort: 9001
+          env:
+          - name: admin_access_log_path
+            value: "{{ .Values.envoyConfiguration.adminAccessLogPath }}"
+          - name: scalardl_address
```

Comment:
One question.
Even scalardl chart and scalardl-audit chart are installed in the same cluster for testing,
does it work properly? (`envoy.fullname` will be different for each dependent chart?)

Reply:
You're right. I think there's something wrong with the value here.
I'll check the operation again.
Thank you very much for your patience. :bow:

Reply:
Usage: 
```
  envoyConfiguration:
    # -- admin log path
    adminAccessLogPath: /dev/stdout
    # envoyConfiguration.upstreamAddress -- upstream address
    upstreamAddress: scalardl-audit-auditor-headless
```

Reply:
Thank you!
So, we need to give the `upstreamAddress` with a custom config every time?
Or is the proper default address (like scalardl-audit-auditor-headless) is automatically set when it's deployed with scalardl-audit chart?

As far as I can see [this](https://github.com/scalar-labs/helm-charts/pull/26/files#diff-a00ee31a72b1a0a7a7b057a256cdd427553e1124cfee2311227ea58580366bd4R8),
it specifies `envoy-headless` so it looks like it won't work with any of the existing charts (like scalardl, scalardl-audit, and scalardb) if it is deployed without a proper custom config.
My understanding is correct?

Reply:
> So, we need to give the upstreamAddress with a custom config every time?
> Or is the proper default address (like scalardl-audit-auditor-headless) is automatically set when it's deployed with scalardl-audit chart?

You need to specify it every time.
Considering the possibility that the service name may be changed by the user, I think it would be better to specify it in Values, so I did it this way.

> it specifies envoy-headless so it looks like it won't work with any of the existing charts (like scalardl, scalardl-audit, and scalardb) if it is deployed without a proper custom config.
> My understanding is correct?

You're right.
I Added mandatory check process.
https://github.com/scalar-labs/helm-charts/pull/26/commits/b1ce6479e1a042e4f922801427fc24a7e082382f#diff-f9e366b1cdbdff67956aecad7953e831aed6c1bc6e0b798f3c5fa59845bb5983R44

https://github.com/scalar-labs/helm-charts/pull/26/commits/b1ce6479e1a042e4f922801427fc24a7e082382f#diff-a00ee31a72b1a0a7a7b057a256cdd427553e1124cfee2311227ea58580366bd4R7-R8

Reply:
I had missed this thread.

The upstream name is dynamically generated as shown below, so it is difficult for the user to specify it.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: {{ include "scalardl.fullname" . }}-ledger-headless
```

- https://github.com/scalar-labs/helm-charts/blob/main/charts/scalardl/templates/ledger/service.yaml#L4

You need to decide on a naming convention and allow it to be determined dynamically:

```yaml
          - name: scalardl_address
            value: {{ include "envoy.fullname" . }}-headless
```

- https://github.com/superbrothers/helm-charts/blob/chart-envoy/charts/envoy/templates/deployment.yaml#L43-L44

You can try adding envoy chart to dependencies of scalardl chart and make sure it works well.

- https://github.com/superbrothers/helm-charts/blob/chart-envoy/charts/scalardl/Chart.yaml#L17-L21

Reply:
I understand that eventually this envoy chart will be added to dependencies of scalardl, scalardb, and scalardl-audit charts. Therefore, users will not explicitly specify the upstream name.

Reply:
Thank you very much. :bow:
I'll check again tomorrow.

Reply:
@superbrothers 
@feeblefakie 
I apologize again for the trouble I've caused you.
The following corrections have been made.
https://github.com/scalar-labs/helm-charts/pull/26/commits/283fe4709271702c8d006f7f62a0ee02877e10cb

The service names for the headless services 'scalardl', 'scalardb', and 'scalardl-audit' are now set as follows
```
apiVersion: v1
kind: Service
metadata:
  name: {{ regexReplaceAll "^(.+)-envoy$" (include "envoy.fullname" .) "$1" }}-headless
```

The environment variable `scalardl_address` of envoy is now set to the same value.
```
- name: scalardl_address
  value: {{ regexReplaceAll "^(.+)-envoy$" (include "envoy.fullname" .) "$1" }}-headless
```

The reason for removing the value envoy is that it is unnatural to have the string envoy in the service names 'scalardl', 'scalardb', and 'scalardl-audit'.

In fact, it was successfully proxied.

```
$ kubectl get deployments scalardb-envoy-envoy -o json | jq '.spec.template.spec.containers[0].env'
[
  {
    "name": "admin_access_log_path",
    "value": "/dev/stdout"
  },
  {
    "name": "scalardl_address",
    "value": "scalardb-headless"
  }
]

kubectl get svc scalardb-headless -o yaml | kubectl neat
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: scalardb
    meta.helm.sh/release-namespace: default
  labels:
    app.kubernetes.io/app: scalardb
    app.kubernetes.io/instance: scalardb
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: scalardb
    app.kubernetes.io/version: 3.1.0
    helm.sh/chart: scalardb-1.0.0
  name: scalardb-headless
  namespace: default
spec:
  clusterIP: None
  ports:
  - name: scalardb
    port: 50051
  selector:
    app.kubernetes.io/app: scalardb
    app.kubernetes.io/instance: scalardb
    app.kubernetes.io/name: scalardb

$ kubectl get deployments scalardl-envoy-envoy -o json | jq '.spec.template.spec.containers[0].env'
[
  {
    "name": "admin_access_log_path",
    "value": "/dev/stdout"
  },
  {
    "name": "scalardl_address",
    "value": "scalardl-headless"
  }
]

$ kubectl get svc scalardl-headless -o yaml | kubectl neat
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: scalardl
    meta.helm.sh/release-namespace: default
  labels:
    app.kubernetes.io/app: ledger
    app.kubernetes.io/instance: scalardl
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: scalardl
    app.kubernetes.io/version: 3.0.1
    helm.sh/chart: scalardl-2.1.0
  name: scalardl-headless
  namespace: default
spec:
  clusterIP: None
  ports:
  - name: scalardl
    port: 50051
  - name: scalardl-admin
    port: 50053
  - name: scalardl-priv
    port: 50052
  selector:
    app.kubernetes.io/app: ledger
    app.kubernetes.io/instance: scalardl
    app.kubernetes.io/name: scalardl

$ kubectl get deployments scalardl-audit-envoy-envoy -o json | jq '.spec.template.spec.containers[0].env'
[
  {
    "name": "admin_access_log_path",
    "value": "/dev/stdout"
  },
  {
    "name": "scalardl_address",
    "value": "scalardl-audit-headless"
  }
]

$ kubectl get svc scalardl-audit-headless -o yaml | kubectl neat
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: scalardl-audit
    meta.helm.sh/release-namespace: default
  labels:
    app.kubernetes.io/app: auditor
    app.kubernetes.io/instance: scalardl-audit
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: scalardl-audit
    app.kubernetes.io/version: 3.0.1
    helm.sh/chart: scalardl-audit-1.0.0
  name: scalardl-audit-headless
  namespace: default
spec:
  clusterIP: None
  ports:
  - name: scalardl-auditor
    port: 50051
  - name: scalardl-auditor-admin
    port: 50053
  - name: scalardl-auditor-priv
    port: 50052
  selector:
    app.kubernetes.io/app: auditor
    app.kubernetes.io/instance: scalardl-audit
    app.kubernetes.io/name: scalardl-audit
```

Reply:
Probably `regexReplaceAll` can be removed, so I will create a patch. Please give me some time.

```yaml
  name: {{ regexReplaceAll "^(.+)-envoy$" (include "envoy.fullname" .) "$1" }}-headless
```

Reply:
You seem to set nameOverride to `*-envoy`, but you can remove `regexpReplaceAll` by simply setting it to `scalardl-ledger` if the chart is `scalardl`.

The below is the case that `nameOverride` is` scalardl-ledger`.

```yaml
# charts/scalardl/values.yaml
envoy:
  enabled: true
  nameOverride: scalardl-ledger
```

You can see that Service name and upstream are the same.

```
$ helm template charts/scalardl | grep -C 4 scalardl-ledger-headless
walk.go:74: found symbolic link in path: /home/ksuda/src/github.com/scalar-labs/helm-charts/charts/scalardl/charts/envoy resolves to /home/ksuda/src/github.com/scalar-labs/helm-charts/charts/envoy
# Source: scalardl/templates/ledger/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: RELEASE-NAME-scalardl-ledger-headless
  labels:
    helm.sh/chart: scalardl-2.1.0
    app.kubernetes.io/name: scalardl
    app.kubernetes.io/instance: RELEASE-NAME
--
          env:
          - name: admin_access_log_path
            value: "/dev/stdout"
          - name: scalardl_address
            value: "RELEASE-NAME-scalardl-ledger-headless"
          livenessProbe:
            httpGet:
              path: /ready
              port: 9001
```

`* -envoy-envoy` service is also improved to `*-envoy`.

```
$ helm template charts/scalardl | grep -C 4 scalardl-ledger-envoy | head -n 10
walk.go:74: found symbolic link in path: /home/ksuda/src/github.com/scalar-labs/helm-charts/charts/scalardl/charts/envoy resolves to /home/ksuda/src/github.com/scalar-labs/helm-charts/charts/envoy
# Source: scalardl/charts/envoy/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: RELEASE-NAME-scalardl-ledger-envoy
  labels:
    helm.sh/chart: envoy-1.0.0
    app.kubernetes.io/name: scalardl-ledger
    app.kubernetes.io/instance: RELEASE-NAME
--
```

The patch is below.

- https://github.com/superbrothers/helm-charts/commit/82320613b0cdc54c8efdbf0afbaa6ab724701baa

If you want Service name to be `scalardl-headless`, you need to modify some template manifests of scalardl, but it is also possible.

Reply:
@superbrothers 
Thank you very much. 🙇 
Only envoy-charts has been fixed.

https://github.com/scalar-labs/helm-charts/pull/26/commits/aa8c473d23839461d18fb398d43d7abea3f63442#diff-f9e366b1cdbdff67956aecad7953e831aed6c1bc6e0b798f3c5fa59845bb5983R44

I'll apply 'nameOverride' to scalardl, scalardb, and scalardl-audit helm-charts after envoy-charts is released.

Reply:
@feeblefakie 
If this is OK, I'll update the `index.yaml` file after the merge and send out the PR for envoy-charts release. :bow: